// cdk8s-cli/lib/import/helm.js
const fs = require('fs');
const path = require('path');
const yaml = require('js-yaml');
const { TypeGenerator } = require('./type-generator');

async function importHelmChart(chartPath, options = {}) {
  const chartYaml = yaml.load(fs.readFileSync(path.join(chartPath, 'Chart.yaml'), 'utf8'));
  const valuesYaml = fs.existsSync(path.join(chartPath, 'values.yaml')) 
    ? yaml.load(fs.readFileSync(path.join(chartPath, 'values.yaml'), 'utf8')) 
    : {};

  const generator = new TypeGenerator({
    definitions: {},
    ...options
  });

  // Process schema definitions with better error handling
  if (fs.existsSync(path.join(chartPath, 'values.schema.json'))) {
    try {
      const schema = JSON.parse(fs.readFileSync(path.join(chartPath, 'values.schema.json'), 'utf8'));
      if (schema.definitions) {
        for (const [name, definition] of Object.entries(schema.definitions)) {
          try {
            generator.addDefinition(name, definition);
          } catch (err) {
            console.warn(`Warning: Skipping invalid schema definition '${name}': ${err.message}`);
          }
        }
      }
      
      // Process root schema with fallback for missing references
      if (schema.properties) {
        for (const [propName, propSchema] of Object.entries(schema.properties)) {
          try {
            generator.addProperty(propName, propSchema);
          } catch (err) {
            console.warn(`Warning: Skipping property '${propName}' due to schema error: ${err.message}`);
            // Add as generic object if schema processing fails
            generator.addProperty(propName, { type: 'object', additionalProperties: true });
          }
        }
      }
    } catch (err) {
      console.warn(`Warning: Failed to process values.schema.json: ${err.message}`);
    }
  }

  // Generate types for values.yaml structure as fallback
  if (Object.keys(valuesYaml).length > 0) {
    generator.addDefinition('Values', {
      type: 'object',
      properties: inferSchemaFromValues(valuesYaml),
      additionalProperties: true
    });
  }

  return {
    chart: chartYaml,
    types: generator.generate()
  };
}

function inferSchemaFromValues(values) {
  const properties = {};
  
  for (const [key, value] of Object.entries(values)) {
    if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
      properties[key] = {
        type: 'object',
        properties: inferSchemaFromValues(value),
        additionalProperties: true
      };
    } else if (Array.isArray(value)) {
      properties[key] = {
        type: 'array',
        items: value.length > 0 ? inferSchemaFromValues({ item: value[0] }).item : { type: 'object' }
      };
    } else {
      properties[key] = {
        type: typeof value
      };
    }
  }
  
  return properties;
}

module.exports = { importHelmChart };
